windows/src/engine/keyman32/preservedkeymap.cpp

```diff
@@ -318,13 +371,114 @@ BOOL PreservedKeyMap::MapKeyboard(KEYBOARD *pKeyboard, PreservedKey **pPreserved
   return TRUE;
 }
 
+BOOL
+PreservedKeyMap::MapKeyboardCore(km_kbp_keyboard *pKeyboard, PreservedKey **pPreservedKeys, size_t *cPreservedKeys) {
+  size_t cKeys = 0, cRules = 0, n = 0;
+  DWORD i;
+
+  m_BaseKeyboardUsesAltGr = KeyboardGivesCtrlRAltForRAlt();  // I4592
+
+  // This is not the same as m_BaseKeyboardUsesAltGr -- we are turning
+  // the simulation back on after (possibly) turning it off, for a
+  // consistent experience. TODO: determine if m_BaseKeyboardUseAltGr
+  // is still needed given we always use kbdus as base for Keyman 10+
+  BOOL bSimulateAltGr = Globals::get_SimulateAltGr();
+
+  // We only want to translate RALT and RALT+SHIFT for Ctrl+Alt rules.
+  // So we exclude all our other favourite modifier keys.
+  const UINT RALT_MATCHING_MASK = TF_MOD_CONTROL | TF_MOD_ALT | TF_MOD_LCONTROL | TF_MOD_RCONTROL | TF_MOD_LALT | TF_MOD_RALT;
+
+  // This is where we will call down to the api to get list of keys used in the keyboard rules
+  km_kbp_keyboard_key *kb_key_list;
+
+  km_kbp_status err_status = km_kbp_keyboard_get_key_list(pKeyboard, &kb_key_list);
+  if ((err_status != KM_KBP_STATUS_OK) || (kb_key_list ==nullptr)) {
+    return FALSE;
+  }
+
+  km_kbp_keyboard_key *key_rule_it = kb_key_list;
+  for (; key_rule_it->key; ++key_rule_it) {
+    ++cRules;
+  }
+  cKeys = cRules;
+  if (cKeys == 0) {
+    km_kbp_keyboard_key_list_dispose(kb_key_list);
+    return FALSE;
+  }
+
+  if (bSimulateAltGr) {
+    // We might need twice as many preserved keys to map both LCtrl+LAlt+x and RAlt+x
+    cKeys *= 2;
+  }
+
+  if (pPreservedKeys == NULL) {
+    *cPreservedKeys = cKeys;
+    km_kbp_keyboard_key_list_dispose(kb_key_list);
+    return TRUE;
+  }
+
+  if (*cPreservedKeys < cKeys) {
```

Comment:
Shouldn't we set `*cPreservedKeys = cKeys` here as well?

Reply:
If we get to this if statement it means the size of the array of PreservedKeys `pPreservedKeys` is might going to be big enough to store all the keys.  So it returns False.

When `::MapKeyboardCore` is called if `pPreservedKeys` is NULL, we set *cPreservedKeys to the size pPreservedKeys  needs to be. 

So in the other cases we don't set `cPreservedKeys`